25962 API reference dokumentation (#114)
magenta-aps/mox: doc/user/objects.rst

           "from": "2017-01-01",
           "to": "2019-03-14"
         }
       }]
     },
     "tilstande": {
       "organisationgyldighed": [{
         "gyldighed": "Aktiv",
         "virkning": {
           "from": "2017-01-01",
           "to": "2019-03-14"
         }
       }
     ]}
   }

The fields used in the example ``organisation`` are the following:

``organisationegenskaber→brugervendtnoegle``
    An ID for the user. Semantically, this should be unique, but we don't
    enforce that. It is intended for the user to recognise the object. Required.

``organisationegenskaber→organisationsnavn``
    The official name of the organisation.

``organisationegenskaber→virkning``
    The period when the above attributes is valid. See :ref:`Valid time`.
    Required.

``organisationgyldighed→gyldighed``
    Whether the organisation is active or not. Can take the values ``Aktiv``
    and ``Inaktiv``. Required.

``organisationgyldighed→virkning``
    The period when the above ``gyldighed`` is valid. See :ref:`Valid time`.
    Required.

---------------------
``organisationenhed``
---------------------

Located at the endpoint :http:get:`!/organisation/organisationenhed`. An

> ``organisationenhed`` is a organisational unit. This could be a department,

Reviewer: "an organisational unit"

.. _Valid time:

----------
Valid time
----------

All attributes and relations have a valid time period associated as
``virkning``. It is the time period during which the fact is true in the real
world.

.. code-block:: json
   :caption: A sample ``virkning`` with required fields
             highlighted.
   :emphasize-lines: 2,5

   {
     "from": "2017-01-01",
     "from_included": true,
     "to": "2025-12-31",
     "to_included": false
   }

The fields used in the example are the following:

``from``
    The time when this fact starts to be true in the real world. Date and time
    input is accepted in almost any reasonable format, including ISO 8601.
    Required.

``from_included``
    Whether the ``from`` timestamp is closed or open. Default ``true``.

``to``
    The time when this facts stop to be true in the real world. Date and time
    input is accepted in almost any reasonable format, including ISO 8601.
    Required.

``to_included``
    Whether the ``to`` timestamp is closed or open. Default ``false``.

.. _transaction time:

----------------
Transaction time
----------------

All transactions also have a transaction time as ``registreret``. This records

> the the time period during which a given fact is stored in the database. With

Reviewer: "the the"

   we actually have more than one version of the protocol to support.

Some ``Organisation`` objects
=============================

----------------
``organisation``
----------------

Located at the endpoint :http:get:`!/organisation/organisation`. An
``organisation`` is a legal organisation. A good example of this a municipality.
The database does support multiple ``organisation``, but in the wild there is
usually only one per MOX instance.

.. code-block:: json
   :caption: A small ``organisation`` with all the required fields highlighted.
   :emphasize-lines: 2-4,6-8,12-17

   {
     "attributter": {
       "organisationegenskaber": [{
         "brugervendtnoegle": "magenta-aps",
         "organisationsnavn": "Magenta ApS",
         "virkning": {
           "from": "2017-01-01",
           "to": "2019-03-14"
         }
       }]
     },
     "tilstande": {
       "organisationgyldighed": [{
         "gyldighed": "Aktiv",
         "virkning": {
           "from": "2017-01-01",
           "to": "2019-03-14"
         }
       }
     ]}
   }

The fields used in the example ``organisation`` are the following:

``organisationegenskaber→brugervendtnoegle``

>     A id for the user. It is not necessarily unique. It is intended for the user

Reviewer 2: ```suggestion
    An id for the user. Semantically, this should be unique, but we don't enforce that. It is intended for the user
```

>     to recognise the object. Required.

Reviewer: "An" id? And ID is an abbreviation and should be in caps.

``organisationegenskaber→organisationsnavn``
    The official name of the organisation.

``organisationegenskaber→virkning``
    The period when the above attributes is valid. See :ref:`Valid time`.
    Required.

``organisationgyldighed→gyldighed``

>     Whether the organisation is active or not. Can takae the values ``Aktiv``

Reviewer 2: ```suggestion
    Whether the organisation is active or not. Can take the values ``Aktiv``
```

    and ``Inaktiv``. Required.

``organisationgyldighed→virkning``
    The period when the above ``gyldighed`` is valid. See :ref:`Valid time`.
    Required.

---------------------
``organisationenhed``
---------------------

> Located at the endpoint :http:get:`!/organisation/organisationenhed`. A

Reviewer 2: ```suggestion
Located at the endpoint :http:get:`!/organisation/organisationenhed`. An
```

> ``organisationenhed`` is a organisational unit. This could be a department,

Reviewer: "an"?

> section, office, committee, project group, class, team and the like. Usually a

Reviewer 2: ```suggestion
section, office, committee, project group, class, team or similar. Usually an
```

``organisation`` contains a single ``organisationenhed`` as a direct decendant
with similar attributes as the parent ``organisation``. This

> ``organisationenhed`` inturn contains all of the organisational heirarchy.

Reviewer 2: ```suggestion
``organisationenhed`` in turn contains all of the organisational heirarchy.
```

Please note that this isn't accurate of how os2mo uses these in practice. We'd _like_ it to be one root, but in practice they have several logical roots.

             highlighted.
   :emphasize-lines: 2-4,6-8,12-17

   {
     "attributter": {
       "organisationenhedegenskaber": [{
         "brugervendtnoegle": "copenhagen",
         "enhedsnavn": "Copenhagen",
         "virkning": {
           "from": "2017-01-01",
           "to": "2019-03-14"
         }
       }]
     },
     "tilstande": {
       "organisationenhedgyldighed": [{
         "gyldighed": "Aktiv",
         "virkning": {
           "from": "2017-01-01",
           "to": "2019-03-14"
         }
       }]
     },
     "relationer": {
       "overordnet": [{
         "uuid": "6ff6cf06-fa47-4bc8-8a0e-7b21763bc30a",
         "virkning": {
           "from": "2017-01-01",
           "to": "2019-03-14"
         }
       }],
       "tilhoerer": [{
         "uuid": "6135c99b-f0fe-4c46-bb50-585b4559b48a",
         "virkning": {
           "from": "2017-01-01",
           "to": "2019-03-14"
         }
       }]
     }
   }

The fields used in the example ``organisationenhed`` are the following:

``organisationenhedegenskaber→*`` and ``organisationenhedgyldighed→*``
   Similar to ``organisation``.

``relationer→tilhoerer``

>    This in the root ``organisation`` which the ``organisationenhed`` is part of.

Reviewer: in->is (?)

   This is usally set on all ``organisationenhed`` to the single
   ``organisation`` in the mox instance.

``relationer→overordnet``
  The parent ``organisationenhed``.

>   On root ``organisationenhed``, it points to the ``organisation``. (This is a

Reviewer: I think this is a part of our use of LoRa in MO and not of LoRa itself. We can choose to point to the organization and we can choose not to. MO chooses to do so, but that's MO-specific and not relevant in LoRa's documentation.

Author: > I think this is a part of our use of LoRa in MO and not of LoRa itself. We can choose to point to the organization and we can choose not to.

I agree.

> MO chooses to do so, but that's MO-specific and not relevant in LoRa's documentation.

I still think it is relevant in the LoRa docs. In the tutorial for the API it is used. How about [`objects.rst#L45`](https://github.com/magenta-aps/mox/pull/114/files/d55f632297d166263b2148cf40aa42928d6af7d5#diff-e547da3bb6cef930e22e11ac059d803aR45) is changed to:
```
The database does support multiple ``organisation``, but combined with `OS2MO
<https://github.com/OS2mo/os2mo>`_ there is usually only one per MOX instance.
```
and this is changed to:
```
With `OS2MO <https://github.com/OS2mo/os2mo>`_ the root ``organisationenhed``
have this field pointing to the ``organisation``. (This is a deliberate
violation of the standard which only allows it to point to
``organisationenhed``.)
```
?

Reviewer 2: > I think this is a part of our use of LoRa in MO and not of LoRa itself. We can choose to point to the organization and we can choose not to. MO chooses to do so, but that's MO-specific and not relevant in LoRa's documentation.

I think we should spell this out, unless we add a way for searching for the _absence_ of a given relation.

Reviewer: In that case (answering Mathias) this is not a documentation of the LoRa API, but an _example_ of a use of it, i.e. in the OS2MO product. We hope to get the opportunity to make many more products with no relations to OS2MO, so maybe this discussion really does belong in the OS2MO documentation.

   :emphasize-lines: 2,5

   {
     "from": "2017-01-01",
     "from_included": true,
     "to": "2025-12-31",
     "to_included": false
   }

The fields used in the example are the following:

``from``
    The time when this facts starts to be true in the real world. Date and time
    input is accepted in almost any reasonable format, including ISO 8601.
    Required.

``from_included``
    Whether the ``from`` timestamp is closed or open. Default ``true``.

``to``
    The time when this facts stop to be true in the real world. Date and time
    input is accepted in almost any reasonable format, including ISO 8601.
    Required.

``to_included``
    Whether the ``to`` timestamp is closed or open. Default ``false``.

.. _transaction time:

----------------
Transaction time
----------------

All transactions also have a transaction time as ``registreret``. This records
the the time period during which a given fact is stored in the database. With
the query parameters to a :ref:`Read <ReadOperation>`, :ref:`List
<ListOperation>` or :ref:`SearchOperation` it can give you a view of the state
of the database at give time in the past.

Integrationdata
===============

> The ``integrationdata``-field is different and its uniqueness is documented here.

Reviewer: As with "virkning" above, no hyphen.
